### PR TITLE
test(melange): demonstrate issue with 2 melange.emit stanzas in the same dir

### PR DESCRIPTION
**Pull request by anmonteiro**
I believe this is a regression introduced within the past 2-3 days

### CODE REVIEW

**Comment by anmonteiro** [comment]
This was broken in https://github.com/ocaml/dune/pull/7211

File: test/blackbox-tests/test-cases/melange/two-emit-stanzas.t [hunk 1]
```diff
@@ -0,0 +1,27 @@
+
+  $ cat > dune-project <<EOF
```
[the remark this answers is not in the document]
Comment (anmonteiro) [reply test/blackbox-tests/test-cases/melange/two-emit-stanzas.t:3]: done

**Comment by anmonteiro** [comment]
fixed the issue in [d647296](https://github.com/ocaml/dune/pull/7229/commits/d647296a9cf21c6046a0f24bd2af6e4ea46d9ebf). The root cause is that we were just returning the first melange.emit stanza in the target directory rather than all.

**Comment by anmonteiro** [comment]
The only test failure is related to #7219 which I'm fixing separately.

File: src/dune_rules/gen_rules.ml [hunk 2]
```diff
@@ -386,35 +386,41 @@ type for_melange =
...
+      List.find_opt for_melange_stanzas ~f:(fun for_melange ->
+          let target_dir =
+            Melange_rules.emit_target_dir ~dir:for_melange.stanza_dir
+              for_melange.stanza
+          in
+          Path.Build.is_descendant dir ~of_:target_dir)
+    with
+    | Some for_melange -> Memo.return (Some for_melange)
+    | None ->
+      let for_melange_in_this_dir = List.hd for_melange_stanzas in
```
Comment (jchavarri) [inline src/dune_rules/gen_rules.ml:422]: The "for_melange" prefixes used across values in this function are a bit confusing to me. What do they mean? Would it make sense to replace them with "emit" or just removed?

**Review by jchavarri — APPROVED** [review]

File: src/dune_rules/gen_rules.ml (hunk 2 above)
Comment (anmonteiro) [reply src/dune_rules/gen_rules.ml:422]: `for_melange` refers to the `type for_melange` just above the diff (you need to expand it once).

File: src/dune_rules/gen_rules.ml [hunk 3]
```diff
@@ -386,35 +386,41 @@ type for_melange =
     stanza : Melange_stanzas.Emit.t
   }
 
-let rec nearest_parent_melange_emit ~dir =
+let rec nearest_parent_melange_emits ~dir =
```
Comment (rgrinberg) [inline src/dune_rules/gen_rules.ml:389]: Can you revert to the original approach with `under_melange_emit_target`? I think it's confusing to collect this list of emit stanzas when only one is going to be relevant.

Comment (anmonteiro) [reply src/dune_rules/gen_rules.ml:389]: I can do that, but I think that'll imply we can't reuse this function.

Comment (rgrinberg) [reply src/dune_rules/gen_rules.ml:389]: I don't see why not, in both uses of this function, you end up looking at only one element in the list.

Comment (anmonteiro) [reply src/dune_rules/gen_rules.ml:389]: yeah, my bad, we can reuse the logic.

Comment (anmonteiro) [reply src/dune_rules/gen_rules.ml]: I think I had just originally overcomplicated the logic here. `a/ (target output)` and `a/b/c/ (target output)` are allowed to co-exist, and I was being overly restrictive. However `a/ (target output)` and `a/output/b` can't, and the original function can detect this case too.

Comment (anmonteiro) [reply src/dune_rules/gen_rules.ml]: This is fixed in the most recent push.

**Pull request merged**